### PR TITLE
Deepbluecapital contribs sql

### PR DESCRIPTION
**Pull request by briot**
This branch contains all contributions proposed by Deep Blue Capital, related to GNATCOLL.SQL.
They provide significant new features and a few bug fixes. They also make the code more extensible by users who want to add their own field types and operators. They also improve efficiency, by as much as 30% in some cases.

The first three commits are already in pull request #12. This is the biggest and hairiest patch, but it has been confirmed as fine already (merge pending update on your side to the infosys testsuite, as I understand).
9d0180a07fb9 is already in pull request #17 

Maintaining 20 local branches (which all depend on one another) does not work. Instead, I suggest we finish integrating #12 and #17, then we can either look at the commits in this list one by one, or do separate pull requests for each of them, depending on what you prefer. I will rebase this branch regularly to keep up to date with AdaCore master.

### CODE REVIEW

**Comment by anisimkov** [comment]
gnatcoll-sql_impl.adb:384:80: (style) this line is too long
gnatcoll-sql_impl.adb:515:80: (style) this line is too long
gnatcoll-sql_impl.adb:1711:05: (style) bad column
gnatcoll-sql_impl.adb:1721:01: (style) trailing spaces not permitted
gnatcoll-sql_impl.adb:1859:05: (style) bad column
gnatcoll-sql_impl.ads:578:05: (style) bad column
gnatcoll-sql_impl.ads:844:42: (style) space required
gnatcoll-sql_impl.ads:846:46: (style) space required
gnatcoll-sql-exec.adb:791:80: (style) this line is too long
gnatcoll-sql-exec.adb:792:80: (style) this line is too long
gnatcoll-sql-exec.adb:802:80: (style) this line is too long

**Comment by anisimkov** [comment]
subtype Criteria_Lists gone from SQL interface. It is incompartibility.

**Comment by briot** [comment]
> subtype Criteria_Lists gone from SQL interface. It is incompartibility.

This type was never meant to be used by users, is not part of the documentation,
and did not even have comments in the code.

**Comment by anisimkov** [comment]
<<is not part of the documentation>> If somebody use it we can't remove it, not depend on documentation. The spec is the root of documentation.

**Comment by briot** [comment]
> <> If somebody use it we can't remove it, not depend on documentation. The spec is the root of documentation.

No, that’s not the rule.

The rule is that we went to be backward compatible whenever it makes sense.
  That means for documented types and subprograms (if undocumented, those are just internal
  details that people depend on).
  Alternatively, if you can provide one use case where this type would have been used, please
  do. I can’t think of any, simply because there are no public subprograms to manipulate this type
  in any case.

**Comment by anisimkov** [comment]
Now I have "raised PROGRAM_ERROR : GNATCOLL.SQL_Impl.Field_List.Implementation.TC_Check: attempt to tamper with cursors" in my project and need GNATCOLL build with debug support to trace the error. Debug mode mean strict style checking, see my prev message about style violations.

**Comment by anisimkov** [comment]
anisimkov@zaria:~/AdaCore/gnatcoll/src> git blame gnatcoll-sql_impl.adb | grep -A 3 "GR.Reserve_Capacity (20)"
f568ee47d (Emmanuel Briot    2017-08-24 11:11:19 +0200  582)                   GR.Reserve_Capacity (20);
f568ee47d (Emmanuel Briot    2017-08-24 11:11:19 +0200  583)                   for C of GR loop
f568ee47d (Emmanuel Briot    2017-08-24 11:11:19 +0200  584)                      GR.Append (C);
f568ee47d (Emmanuel Briot    2017-08-24 11:11:19 +0200  585)                   end loop;

It is not right to append a container in a loop over the same container.

**Comment by briot** [comment]
> anisimkov@zaria:~/AdaCore/gnatcoll/src> git blame gnatcoll-sql_impl.adb | grep -A 3 "GR.Reserve_Capacity (20)"
> f568ee4 (Emmanuel Briot 2017-08-24 11:11:19 +0200 582) GR.Reserve_Capacity (20);
> f568ee4 (Emmanuel Briot 2017-08-24 11:11:19 +0200 583) for C of GR loop
> f568ee4 (Emmanuel Briot 2017-08-24 11:11:19 +0200 584) GR.Append (C);
> f568ee4 (Emmanuel Briot 2017-08-24 11:11:19 +0200 585) end loop;
> 
> It is not right to append a container in a loop over the same container.


Well spotted !

Fixed on github by amending the corresponding commit

**Comment by anisimkov** [comment]
diff --git a/src/gnatcoll-sql_impl.adb b/src/gnatcoll-sql_impl.adb
index 28f41743..a8df03a6 100644
--- a/src/gnatcoll-sql_impl.adb
+++ b/src/gnatcoll-sql_impl.adb
@@ -508,9 +508,6 @@ package body GNATCOLL.SQL_Impl is
    begin
       if Left.List.Is_Null then
          return +Right;
-      elsif Left.List.Get_Refcount = 1 then
-         Left.List.Get.Append (Right);  --  Modify in place
-         return Left;
       else
          return Result : SQL_Field_List do
             Result.List.Set (Left.List.Get); --  Copy the list to preserve Left

You can't modify in place, because in case of list is declared constant and used to add fields sometimes, you are going to increase it in place.

**Comment by briot** [comment]
> 
> You can't modify in place, because in case of list is declared constant and used to add fields sometimes, you are going to increase it in place.


Agreed. For some reason, I thought I had tested and that the test on the refcount
was enough to avoid that, but it isn’t.
I have amended the commit to remove the invalid optimization, and added an
automatic test.

Thanks

**Comment by anisimkov** [comment]
```
package body GNATCOLL.SQL is
...
   function Combine
     (Left, Right : SQL_Criteria; Op : Criteria_Combine) return SQL_Criteria
...
         if not Is_Shared (Left) then
            Append_Right_To_List (Local_Criteria (LD).Criterias);
            return Left;

         else
```
Should not copy in place this way too.

**Comment by anisimkov** [comment]
By the way, can you explain, why does it mean "shared" ?
```
   function Is_Shared (Self : SQL_Criteria) return Boolean is
   begin
      return Self.Criteria.Get_Refcount = 1;
   end Is_Shared;
```
The only one ref mean "not shared" as I can understand.

**Comment by briot** [comment]
>          if not Is_Shared (Left) then
>             Append_Right_To_List (Local_Criteria (LD).Criterias);
>             return Left;
> 
>          else
> 
> Should not copy in place this way too.

I left this one because I could not make it fail (see the automatic test that
was added)

**Comment by anisimkov** [comment]
<< I could not make it fail >> I guess it is just because of 2 errors compensating one another.

**Comment by briot** [comment]
> << I could not make it fail >> I guess it is just because of 2 errors compensating one another.

Which errors ?
The test does:

      declare
         C1 : constant SQL_Criteria := Config.Sales_Message = True;
         C2 : constant SQL_Criteria := C1 and T1.What_Done = "";
      begin
         Assert (TS (C1, Sqlite), "config.sales_message=1");
         Assert
            (TS (C2, Sqlite), "config.sales_message=1 AND foo.what_done=''”);
      end;

So C1 would be modified when we create C2 if there was an error. But it
isn’t, so I think all is fine.

**Comment by anisimkov** [comment]
Ok, how about question about "shared" ?

**Comment by briot** [comment]
> Ok, how about question about "shared" ?

Whether refcount > 1 (i.e. shared among multiple variables, though some of those might be temporaries generated by the compiler)

**Comment by anisimkov** [comment]
Is there ">" ?
```
   function Is_Shared (Self : SQL_Criteria) return Boolean is
   begin
      return Self.Criteria.Get_Refcount = 1;
   end Is_Shared;
```

**Comment by briot** [comment]
> Is there ">" ?
> 
>    function Is_Shared (Self : SQL_Criteria) return Boolean is
>    begin
>       return Self.Criteria.Get_Refcount = 1;
>    end Is_Shared;


OK, now I understand the question. And indeed we should be using Not_Shared,
or “>1”.

**Comment by briot** [comment]
> OK, now I understand the question. And indeed we should be using Not_Shared,
> or “>1”.

So I performed more testing, and in fact the “if not Is_Shared” test was never
true (as expected since you pointed out Is_Shared is badly implemented).
To be safe, I simply decided to remove Is_Shared altogether.

**Comment by anisimkov** [comment]
Manu, 

As I already wrote we had to revert this patch because of GNAT error on 32 bit platforms. We are investigating the error. Besides that I'm not sure that it is GNAT problem, because when I changed XString to Unbounded_String in Strings_Imps (using renames) the error is disappeared. And as I can see you use System.Word_Size in the Strings implementation which is different in 32 and 64 bit platforms. I think you can try to compile your code on 32 bit platform and maybe you will have an idea.

**Comment by briot** [comment]
> As I already wrote we had to revert this patch because of GNAT error on 32 bit platforms. We are investigating the error. Besides that I'm not sure that it is GNAT problem, because when I changed XString to Unbounded_String in Strings_Imps (using renames) the error is disappeared. And as I can see you use System.Word_Size in the Strings implementation which is different in 32 and 64 bit platforms. I think you can try to compile your code on 32 bit platform and maybe you will have an idea.


I do not have this compiler, nor any reason to install one, really. That’s not a compiler
we use.

Thanks for the update though

File: src/postgres/gnatcoll-sql-postgres.ads [hunk 1]
```diff
@@ -49,6 +49,19 @@ package GNATCOLL.SQL.Postgres is
    --    Prefer   => first try a SSL connection, then non-SSL if failed
    --    Require  => require a SSL connection
 
+   type Pgbouncer_Config is
```
Comment (anisimkov) [inline src/postgres/gnatcoll-sql-postgres.ads]: As I can remember we decided to do not note "Pgbouncer" but just note disable "prepare on server".

**Comment by briot** [comment]
> In src/postgres/gnatcoll-sql-postgres.ads:
> 
> > @@ -49,6 +49,19 @@ package GNATCOLL.SQL.Postgres is
>     --    Prefer   => first try a SSL connection, then non-SSL if failed
>     --    Require  => require a SSL connection
>  
> +   type Pgbouncer_Config is
> 
> As I can remember we decided to do not note "Pgbouncer" but just note disable "prepare on server".


Dmitry,

That’s correct, and indeed I have on my list to do this change. Haven’t got around to it
yet, but this is pretty easy to do so won’t take very long.
Right now I can’t really merge things into gnatcoll since you had to back out the sql
patch, so I’ll wait for that one and then contribute some more changes including the
removal of pgbouncer_config

Thanks

Manu

**Comment by anisimkov** [comment]
One more problem with this patch. Regrests failed when portgress support turned off. gnatcoll-db2ada
tryting to create postgress db connection even when it is no necessary.

**Comment by setton** [comment]
Just to give a heads-up to everyone: we have encountered a lot of problems with a previous GNATCOLL patch (affecting the behavior of GNATCOLL.Traces, making traces activated that were previously deactivated) so we might need to revert, which of course might impact our ability to process new patches. Also we're internally about to enter "freeze mode", so we might wait until the 18 branch is created before we can include anything that's big and potentially impacting.

**Comment by briot** [comment]
> Just to give a heads-up to everyone: we have encountered a lot of problems with a previous GNATCOLL patch (affecting the behavior of GNATCOLL.Traces, making traces activated that were previously deactivated)

If this is the issue with “NAME = yes”, then a pull request has been made already

**Comment by briot** [comment]
> One more problem with this patch. Regrests failed when portgress support turned off. gnatcoll-db2ada
> tryting to create postgress db connection even when it is no necessary.

I don’t think there’s any change related to any recent patch of mine.
gnatcoll_db2ada supports multiple backends, and the command line can be used
to select one or the other. If we call it with no specific database, then it will indeed
try to connect to postgres (and has for the last 12 - 15 years).

Perhaps some tests need to be disabled when postgres is disabled (with a
decorator in __init__.py)

**Comment by anisimkov** [comment]
Just try to ./configure --without-postgresql and you'll get a few regressions on sql testuites.
Without this patch everithing ok.

**Comment by anisimkov** [comment]
even on those testsuite which are only generating texts and does not need any databases.

**Comment by briot** [comment]
> Just try to ./configure --without-postgresql and you'll get a few regressions on sql testuites.
> Without this patch everithing ok.

I’ll take a look when I am again able to merge things (which is not the case since you
have reverted my patches). So like the pgbouncer, I’ll wait for now until the GNAT
issue is fixed.

**Comment by anisimkov** [comment]
GNAT 32 bit issue fixed in recent wavefronts.
The merge problem not because of revert, but because on fix old error in 
2fc97da391839eb0dd49aed2550440111a0666ae.

**Pull request merged**